pkg/ipmi/oem.go

```diff
@@ -5,16 +5,299 @@
+	ParameterSelector       uint8
+	DIMMPresent             uint8
+	NodeNumber              uint8
+	ChannelNumber           uint8
+	DIMMNumber              uint8
+	DIMMType                uint8
+	DIMMSpeed               uint16
+	DIMMSize                uint32
+	ModulePartNumber        [20]byte
+	ModuleSerialNumber      uint32
+	ModuleManufacturerIDLSB uint8
+	ModuleManufacturerIDMSB uint8
+}
+
+var BoardManufacturer = map[string][3]uint8{
+	"Wiwynn": {0x0, 0x9c, 0x9c},
+}
+
+func (i *IPMI) SendOemIpmiProcessorInfo(info *ProcessorInfo) error {
+	req := &req{}
+	req.msg.cmd = _FB_OEM_SET_PROC_INFO
+	req.msg.netfn = _IPMI_FB_OEM_NET_FUNCTION2
+
+	for param := 1; param <= 2; param++ {
+		data, err := info.marshall(param)
+		if err != nil {
+			return err
+		}
+
+		req.msg.data = unsafe.Pointer(&data[0])
+		req.msg.dataLen = uint16(len(data))
+
+		_, err = i.sendrecv(req)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (i *IPMI) SendOemIpmiDimmInfo(info *DimmInfo) error {
+	req := &req{}
+	req.msg.cmd = _FB_OEM_SET_DIMM_INFO
+	req.msg.netfn = _IPMI_FB_OEM_NET_FUNCTION2
+
+	for param := 1; param <= 6; param++ {
+		//If DIMM is not present, only send the information of DIMM location
+		if info.DIMMPresent != 0x01 && param >= 2 {
+			continue
+		}
+
+		data, err := info.marshall(param)
+		if err != nil {
+			return err
+		}
+
+		req.msg.data = unsafe.Pointer(&data[0])
+		req.msg.dataLen = uint16(len(data))
+
+		_, err = i.sendrecv(req)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *ProcessorInfo) marshall(param int) ([]byte, error) {
+	var data []byte
+	buf := &bytes.Buffer{}
+
+	if err := binary.Write(buf, binary.LittleEndian, *p); err != nil {
+		return nil, err
+	}
+
+	buf.Bytes()[4] = byte(param)
+
+	switch param {
+	case 1:
+		data = make([]byte, 53)
+		copy(data[:], buf.Bytes()[:53])
+	case 2:
+		data = make([]byte, 12)
+		copy(data[0:5], buf.Bytes()[0:5])
+		copy(data[5:12], buf.Bytes()[53:60])
+	}
+
+	return data, nil
+}
+
+func (d *DimmInfo) marshall(param int) ([]byte, error) {
+	var data []byte
+	buf := &bytes.Buffer{}
+
+	if err := binary.Write(buf, binary.LittleEndian, *d); err != nil {
+		return nil, err
+	}
+
+	buf.Bytes()[4] = byte(param)
+
+	switch param {
+	case 1:
+		data = make([]byte, 9)
+		copy(data[:], buf.Bytes()[:9])
+	case 2:
+		data = make([]byte, 6)
+		copy(data[0:5], buf.Bytes()[0:5])
+		copy(data[5:6], buf.Bytes()[9:10])
+	case 3:
+		data = make([]byte, 11)
+		copy(data[0:5], buf.Bytes()[0:5])
+		copy(data[5:11], buf.Bytes()[10:16])
+	case 4:
+		data = make([]byte, 25)
+		copy(data[0:5], buf.Bytes()[0:5])
+		copy(data[5:25], buf.Bytes()[16:36])
+	case 5:
+		data = make([]byte, 9)
+		copy(data[0:5], buf.Bytes()[0:5])
+		copy(data[5:9], buf.Bytes()[36:40])
+	case 6:
+		data = make([]byte, 7)
+		copy(data[0:5], buf.Bytes()[0:5])
+		copy(data[5:7], buf.Bytes()[40:42])
+	}
+
+	return data, nil
+}
+
+func (i *IPMI) IpmiOemSetProcessorInfo(si *smbios.Info) error {
```

Comment:
Diff looks mostly good to me.

The only thing I am not totally convinced about is having the IPMI library filling the Dimm and Processor structs. I think the IPMI library should only focus on data exchange and not filling in the actual data that needs to be sent. I also think it should not depend on the `smbios` package.

However, maybe it is just me. If you think this is the best place to collect that data feel free to leave it there.

I am open to suggestions.

Reply:
I implemented a function for collecting the data and called in the function using the IPMI library.

Reply:
Back to @pmazzini 's point, maybe move the filling the Dimm and Processor functions to systemboot/oem.go? Like this commit for example:
https://github.com/johnnylinwiwynn/u-root/commit/3d497c1fa1885493e6382d620f34b08ae015145b

Reply:
@pmazzini How about Johnny`s suggestion?